File: python/src/exercises/task06/string_manipulation_exercises.py
"""
Task 6: String Manipulation Exercises

This module contains exercises for learning fundamental string manipulation operations in Python.
Students will implement methods that demonstrate:
- Basic string operations (length, indexing, slicing)
- String comparison and searching
- String modification (replace, upper, lower)
- String splitting and joining
- String validation and checking
- Advanced string manipulation techniques

Learning Objectives:
- Master Python string methods and operations
- Understand string immutability in Python
- Practice string validation and parsing
- Learn Pythonic string manipulation patterns
- Handle edge cases with None and empty strings

Author: ZipCode Cohort
"""

from typing import List, Optional, Union
import re


class StringManipulationExercises:
    """
    String manipulation exercises for learning Python string operations.
    """

    def get_string_length(self, text: Optional[str]) -> int:
        """
        Get the length of a string.
        Handle None strings by returning 0.
        
        Examples:
        >>> exercises = StringManipulationExercises()
        >>> exercises.get_string_length('hello')
        5
        >>> exercises.get_string_length('')
        0
        >>> exercises.get_string_length(None)
        0
        
        Args:
            text: Input string
            
        Returns:
            Length of the string, 0 if None
        """
        raise NotImplementedError("Implement get_string_length")

    def to_upper_case(self, text: Optional[str]) -> Optional[str]:
        """
        Convert a string to uppercase.
        Handle None strings by returning None.
        
        Examples:
        >>> exercises = StringManipulationExercises()
        >>> exercises.to_upper_case('hello')
        'HELLO'
        >>> exercises.to_upper_case('Hello World')
        'HELLO WORLD'
        >>> exercises.to_upper_case(None)
        None
        
        Args:
            text: Input string
            
        Returns:
            Uppercase version of the string
        """
        raise NotImplementedError("Implement to_upper_case")

    def to_lower_case(self, text: Optional[str]) -> Optional[str]:
        """
        Convert a string to lowercase.
        Handle None strings by returning None.
        
        Examples:
        >>> exercises = StringManipulationExercises()
        >>> exercises.to_lower_case('HELLO')
        'hello'
        >>> exercises.to_lower_case('Hello World')
        'hello world'
        >>> exercises.to_lower_case(None)
        None
        
        Args:
            text: Input string
            
        Returns:
            Lowercase version of the string
        """
        raise NotImplementedError("Implement to_lower_case")

    def contains_substring(self, text: Optional[str], substring: Optional[str]) -> bool:
        """
        Check if a string contains a substring (case-sensitive).
        Return False if either string is None.
        
        Examples:
        >>> exercises = StringManipulationExercises()
        >>> exercises.contains_substring('hello world', 'world')
        True
        >>> exercises.contains_substring('hello world', 'WORLD')
        False
        >>> exercises.contains_substring('hello world', 'xyz')
        False
        >>> exercises.contains_substring(None, 'test')
        False
        
        Args:
            text: Main string to search in
            substring: Substring to search for
            
        Returns:
            True if substring is found, False otherwise
        """
        raise NotImplementedError("Implement contains_substring")

    def get_substring(self, text: Optional[str], start: int, end: int) -> Optional[str]:
        """
        Get a substring from start index to end index (exclusive).
        Handle out-of-bounds indices gracefully.
        
        Examples:
        >>> exercises = StringManipulationExercises()
        >>> exercises.get_substring('hello', 1, 4)
        'ell'
        >>> exercises.get_substring('hello', 0, 5)
        'hello'
        >>> exercises.get_substring('hello', 2, 2)
        ''
        >>> exercises.get_substring('hello', -1, 3)
        'hel'
        
        Args:
            text: Input string
            start: Start index (inclusive)
            end: End index (exclusive)
            
        Returns:
            Substring from start to end
        """
        raise NotImplementedError("Implement get_substring")

    def replace_char(self, text: Optional[str], old_char: str, new_char: str) -> Optional[str]:
        """
        Replace all occurrences of old_char with new_char.
        Handle None strings by returning None.
        
        Examples:
        >>> exercises = StringManipulationExercises()
        >>> exercises.replace_char('hello', 'l', 'x')
        'hexxo'
        >>> exercises.replace_char('programming', 'm', 'M')
        'prograMMing'
        >>> exercises.replace_char('test', 'z', 'a')
        'test'
        
        Args:
            text: Input string
            old_char: Character to replace
            new_char: Character to replace with
            
        Returns:
            String with characters replaced
        """
        raise NotImplementedError("Implement replace_char")

    def replace_substring(self, text: Optional[str], old_substring: str, new_substring: str) -> Optional[str]:
        """
        Replace all occurrences of a substring with another substring.
        Handle None strings by returning None.
        
        Examples:
        >>> exercises = StringManipulationExercises()
        >>> exercises.replace_substring('hello world', 'world', 'Python')
        'hello Python'
        >>> exercises.replace_substring('aaa', 'aa', 'b')
        'ba'
        >>> exercises.replace_substring('test', 'xyz', 'abc')
        'test'
        
        Args:
            text: Input string
            old_substring: Substring to replace
            new_substring: Substring to replace with
            
        Returns:
            String with substrings replaced
        """
        raise NotImplementedError("Implement replace_substring")

    def split_string(self, text: Optional[str], delimiter: str) -> Optional[List[str]]:
        """
        Split a string by a delimiter into a list of strings.
        Handle None or empty strings appropriately.
        
        Examples:
        >>> exercises = StringManipulationExercises()
        >>> exercises.split_string('apple,banana,cherry', ',')
        ['apple', 'banana', 'cherry']
        >>> exercises.split_string('one-two-three', '-')
        ['one', 'two', 'three']
        >>> exercises.split_string('noseparator', ',')
        ['noseparator']
        >>> exercises.split_string('', ',')
        ['']
        
        Args:
            text: Input string to split
            delimiter: Delimiter to split by
            
        Returns:
            List of split strings
        """
        raise NotImplementedError("Implement split_string")

    def join_strings(self, strings: Optional[List[str]], separator: str) -> Optional[str]:
        """
        Join a list of strings with a separator.
        Handle None lists by returning None.
        Handle None strings in list by treating them as empty strings.
        
        Examples:
        >>> exercises = StringManipulationExercises()
        >>> exercises.join_strings(['apple', 'banana', 'cherry'], ',')
        'apple,banana,cherry'
        >>> exercises.join_strings(['one', 'two', 'three'], '-')
        'one-two-three'
        >>> exercises.join_strings(['single'], ',')
        'single'
        >>> exercises.join_strings([], ',')
        ''
        
        Args:
            strings: List of strings to join
            separator: Separator to use between strings
            
        Returns:
            Joined string
        """
        raise NotImplementedError("Implement join_strings")

    def trim_string(self, text: Optional[str]) -> Optional[str]:
        """
        Remove whitespace from the beginning and end of a string.
        Handle None strings by returning None.
        
        Examples:
        >>> exercises = StringManipulationExercises()
        >>> exercises.trim_string('  hello world  ')
        'hello world'
        >>> exercises.trim_string('\\t\\nhello\\n\\t')
        'hello'
        >>> exercises.trim_string('no spaces')
        'no spaces'
        >>> exercises.trim_string('   ')
        ''
        
        Args:
            text: Input string to trim
            
        Returns:
            Trimmed string
        """
        raise NotImplementedError("Implement trim_string")

    def is_empty_or_whitespace(self, text: Optional[str]) -> bool:
        """
        Check if a string is None, empty, or contains only whitespace.
        Return True for None strings.
        
        Examples:
        >>> exercises = StringManipulationExercises()
        >>> exercises.is_empty_or_whitespace('')
        True
        >>> exercises.is_empty_or_whitespace('   ')
        True
        >>> exercises.is_empty_or_whitespace('\\t\\n')
        True
        >>> exercises.is_empty_or_whitespace('hello')
        False
        >>> exercises.is_empty_or_whitespace(None)
        True
        
        Args:
            text: Input string to check
            
        Returns:
            True if string is None, empty, or whitespace only
        """
        raise NotImplementedError("Implement is_empty_or_whitespace")

    def count_occurrences(self, text: Optional[str], char: str) -> int:
        """
        Count the number of occurrences of a character in a string.
        Handle None strings by returning 0.
        
        Examples:
        >>> exercises = StringManipulationExercises()
        >>> exercises.count_occurrences('hello', 'l')
        2
        >>> exercises.count_occurrences('programming', 'g')
        2
        >>> exercises.count_occurrences('test', 'z')
        0
        >>> exercises.count_occurrences('', 'a')
        0
        
        Args:
            text: Input string to search
            char: Character to count
            
        Returns:
            Number of occurrences of the character
        """
        raise NotImplementedError("Implement count_occurrences")

    def reverse_string(self, text: Optional[str]) -> Optional[str]:
        """
        Reverse a string.
        Handle None strings by returning None.
        
        Examples:
        >>> exercises = StringManipulationExercises()
        >>> exercises.reverse_string('hello')
        'olleh'
        >>> exercises.reverse_string('Python')
        'nohtyP'
        >>> exercises.reverse_string('')
        ''
        >>> exercises.reverse_string('a')
        'a'
        
        Args:
            text: Input string to reverse
            
        Returns:
            Reversed string
        """
        raise NotImplementedError("Implement reverse_string")

    def is_palindrome(self, text: Optional[str]) -> bool:
        """
        Check if a string is a palindrome (reads the same forwards and backwards).
        Case-insensitive comparison, ignoring spaces and punctuation.
        Handle None strings by returning False.
        
        Examples:
        >>> exercises = StringManipulationExercises()
        >>> exercises.is_palindrome('racecar')
        True
        >>> exercises.is_palindrome('A man a plan a canal Panama')
        True
        >>> exercises.is_palindrome('hello')
        False
        >>> exercises.is_palindrome('')
        True
        
        Args:
            text: Input string to check
            
        Returns:
            True if string is a palindrome
        """
        raise NotImplementedError("Implement is_palindrome")

    def capitalize_words(self, text: Optional[str]) -> Optional[str]:
        """
        Capitalize the first letter of each word in a string.
        Words are separated by spaces.
        Handle None strings by returning None.
        
        Examples:
        >>> exercises = StringManipulationExercises()
        >>> exercises.capitalize_words('hello world')
        'Hello World'
        >>> exercises.capitalize_words('python programming')
        'Python Programming'
        >>> exercises.capitalize_words('a')
        'A'
        >>> exercises.capitalize_words('')
        ''
        
        Args:
            text: Input string to capitalize
            
        Returns:
            String with first letter of each word capitalized
        """
        raise NotImplementedError("Implement capitalize_words")

    def extract_words(self, text: Optional[str]) -> List[str]:
        """
        Extract all words from a string (separated by spaces).
        Remove empty strings from the result.
        Handle multiple consecutive spaces.
        
        Examples:
        >>> exercises = StringManipulationExercises()
        >>> exercises.extract_words('hello world python')
        ['hello', 'world', 'python']
        >>> exercises.extract_words('  multiple   spaces  ')
        ['multiple', 'spaces']
        >>> exercises.extract_words('')
        []
        >>> exercises.extract_words('single')
        ['single']
        
        Args:
            text: Input string to extract words from
            
        Returns:
            List of words extracted from the string
        """
        raise NotImplementedError("Implement extract_words")

    def is_valid_integer(self, text: Optional[str]) -> bool:
        """
        Check if a string represents a valid integer.
        Handle negative numbers and leading/trailing whitespace.
        
        Examples:
        >>> exercises = StringManipulationExercises()
        >>> exercises.is_valid_integer('123')
        True
        >>> exercises.is_valid_integer('-456')
        True
        >>> exercises.is_valid_integer('  789  ')
        True
        >>> exercises.is_valid_integer('12.34')
        False
        >>> exercises.is_valid_integer('abc')
        False
        >>> exercises.is_valid_integer('')
        False
        
        Args:
            text: String to check
            
        Returns:
            True if string represents a valid integer
        """
        raise NotImplementedError("Implement is_valid_integer")

    def remove_non_alphabetic(self, text: Optional[str]) -> Optional[str]:
        """
        Remove all non-alphabetic characters from a string.
        Keep only letters (a-z, A-Z).
        
        Examples:
        >>> exercises = StringManipulationExercises()
        >>> exercises.remove_non_alphabetic('Hello123World!')
        'HelloWorld'
        >>> exercises.remove_non_alphabetic('Python2024')
        'Python'
        >>> exercises.remove_non_alphabetic('123!@#')
        ''
        >>> exercises.remove_non_alphabetic('')
        ''
        
        Args:
            text: Input string to clean
            
        Returns:
            String with only alphabetic characters
        """
        raise NotImplementedError("Implement remove_non_alphabetic")

    def string_format_example(self, name: str, age: int, score: float) -> str:
        """
        Demonstrate Python string formatting using f-strings.
        Create a formatted message with the given parameters.
        
        Examples:
        >>> exercises = StringManipulationExercises()
        >>> exercises.string_format_example('Alice', 25, 95.5)
        'Name: Alice, Age: 25, Score: 95.50'
        
        Args:
            name: Person's name
            age: Person's age
            score: Person's score
            
        Returns:
            Formatted string
        """
        raise NotImplementedError("Implement string_format_example")

    def find_all_indices(self, text: Optional[str], substring: str) -> List[int]:
        """
        Find all starting indices where substring appears in text.
        Return empty list if no matches found or text is None.
        
        Examples:
        >>> exercises = StringManipulationExercises()
        >>> exercises.find_all_indices('hello hello world', 'hello')
        [0, 6]
        >>> exercises.find_all_indices('abababab', 'ab')
        [0, 2, 4, 6]
        >>> exercises.find_all_indices('test', 'xyz')
        []
        
        Args:
            text: Text to search in
            substring: Substring to find
            
        Returns:
            List of starting indices where substring is found
        """
        raise NotImplementedError("Implement find_all_indices")

    def longest_word(self, text: Optional[str]) -> Optional[str]:
        """
        Find the longest word in a string.
        If multiple words have the same length, return the first one.
        Words are separated by spaces.
        
        Examples:
        >>> exercises = StringManipulationExercises()
        >>> exercises.longest_word('hello world python programming')
        'programming'
        >>> exercises.longest_word('cat dog elephant')
        'elephant'
        >>> exercises.longest_word('a bb cc')
        'bb'
        >>> exercises.longest_word('')
        None
        
        Args:
            text: Input string to search
            
        Returns:
            Longest word, or None if no words found
        """
        raise NotImplementedError("Implement longest_word")

File: java/src/main/java/com/zipcode/exercises/task06/StringManipulationExercises.java
package com.zipcode.exercises.task06;

import java.util.List;

/**
 * Task 6: String Manipulation Exercises
 * 
 * This class contains exercises for learning fundamental string manipulation operations in Java.
 * Students will implement methods that demonstrate:
 * - Basic string operations (length, indexing, substring)
 * - String comparison and searching
 * - String modification (replace, uppercase, lowercase)
 * - String splitting and joining
 * - StringBuilder usage for efficient string building
 * - String validation and checking
 * 
 * Learning Objectives:
 * - Understand immutability of String objects in Java
 * - Master common String class methods
 * - Learn efficient string building with StringBuilder
 * - Practice string validation and parsing
 * - Handle edge cases with null and empty strings
 * 
 * @author ZipCode Cohort
 */
public class StringManipulationExercises {

    /**
     * Get the length of a string.
     * Handle null strings by returning 0.
     * 
     * Examples:
     * getStringLength("hello") → 5
     * getStringLength("") → 0
     * getStringLength(null) → 0
     * 
     * @param str Input string
     * @return Length of the string, 0 if null
     */
    public int getStringLength(String str) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Convert a string to uppercase.
     * Handle null strings by returning null.
     * 
     * Examples:
     * toUpperCase("hello") → "HELLO"
     * toUpperCase("Hello World") → "HELLO WORLD"
     * toUpperCase(null) → null
     * 
     * @param str Input string
     * @return Uppercase version of the string
     */
    public String toUpperCase(String str) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Convert a string to lowercase.
     * Handle null strings by returning null.
     * 
     * Examples:
     * toLowerCase("HELLO") → "hello"
     * toLowerCase("Hello World") → "hello world"
     * toLowerCase(null) → null
     * 
     * @param str Input string
     * @return Lowercase version of the string
     */
    public String toLowerCase(String str) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Check if a string contains a substring (case-sensitive).
     * Return false if either string is null.
     * 
     * Examples:
     * containsSubstring("hello world", "world") → true
     * containsSubstring("hello world", "WORLD") → false
     * containsSubstring("hello world", "xyz") → false
     * containsSubstring(null, "test") → false
     * 
     * @param str Main string to search in
     * @param substring Substring to search for
     * @return True if substring is found, false otherwise
     */
    public boolean containsSubstring(String str, String substring) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Get a substring from start index to end index (exclusive).
     * Handle out-of-bounds indices gracefully.
     * 
     * Examples:
     * getSubstring("hello", 1, 4) → "ell"
     * getSubstring("hello", 0, 5) → "hello"
     * getSubstring("hello", 2, 2) → ""
     * getSubstring("hello", -1, 3) → "hel" (treat negative as 0)
     * 
     * @param str Input string
     * @param start Start index (inclusive)
     * @param end End index (exclusive)
     * @return Substring from start to end
     */
    public String getSubstring(String str, int start, int end) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Replace all occurrences of oldChar with newChar.
     * Handle null strings by returning null.
     * 
     * Examples:
     * replaceChar("hello", 'l', 'x') → "hexxo"
     * replaceChar("programming", 'm', 'M') → "prograMMing"
     * replaceChar("test", 'z', 'a') → "test"
     * 
     * @param str Input string
     * @param oldChar Character to replace
     * @param newChar Character to replace with
     * @return String with characters replaced
     */
    public String replaceChar(String str, char oldChar, char newChar) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Replace all occurrences of a substring with another substring.
     * Handle null strings by returning null.
     * 
     * Examples:
     * replaceSubstring("hello world", "world", "Java") → "hello Java"
     * replaceSubstring("aaa", "aa", "b") → "ba"
     * replaceSubstring("test", "xyz", "abc") → "test"
     * 
     * @param str Input string
     * @param oldSubstring Substring to replace
     * @param newSubstring Substring to replace with
     * @return String with substrings replaced
     */
    public String replaceSubstring(String str, String oldSubstring, String newSubstring) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Split a string by a delimiter into an array of strings.
     * Handle null or empty strings appropriately.
     * 
     * Examples:
     * splitString("apple,banana,cherry", ",") → ["apple", "banana", "cherry"]
     * splitString("one-two-three", "-") → ["one", "two", "three"]
     * splitString("noseparator", ",") → ["noseparator"]
     * splitString("", ",") → [""]
     * 
     * @param str Input string to split
     * @param delimiter Delimiter to split by
     * @return Array of split strings
     */
    public String[] splitString(String str, String delimiter) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Join an array of strings with a separator.
     * Handle null arrays by returning null.
     * Handle null strings in array by treating them as empty strings.
     * 
     * Examples:
     * joinStrings(["apple", "banana", "cherry"], ",") → "apple,banana,cherry"
     * joinStrings(["one", "two", "three"], "-") → "one-two-three"
     * joinStrings(["single"], ",") → "single"
     * joinStrings([], ",") → ""
     * 
     * @param strings Array of strings to join
     * @param separator Separator to use between strings
     * @return Joined string
     */
    public String joinStrings(String[] strings, String separator) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Remove whitespace from the beginning and end of a string.
     * Handle null strings by returning null.
     * 
     * Examples:
     * trimString("  hello world  ") → "hello world"
     * trimString("\t\nhello\n\t") → "hello"
     * trimString("no spaces") → "no spaces"
     * trimString("   ") → ""
     * 
     * @param str Input string to trim
     * @return Trimmed string
     */
    public String trimString(String str) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Check if a string is empty or contains only whitespace.
     * Return true for null strings.
     * 
     * Examples:
     * isEmptyOrWhitespace("") → true
     * isEmptyOrWhitespace("   ") → true
     * isEmptyOrWhitespace("\t\n") → true
     * isEmptyOrWhitespace("hello") → false
     * isEmptyOrWhitespace(null) → true
     * 
     * @param str Input string to check
     * @return True if string is empty or whitespace only
     */
    public boolean isEmptyOrWhitespace(String str) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Count the number of occurrences of a character in a string.
     * Handle null strings by returning 0.
     * 
     * Examples:
     * countOccurrences("hello", 'l') → 2
     * countOccurrences("programming", 'g') → 2
     * countOccurrences("test", 'z') → 0
     * countOccurrences("", 'a') → 0
     * 
     * @param str Input string to search
     * @param ch Character to count
     * @return Number of occurrences of the character
     */
    public int countOccurrences(String str, char ch) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Reverse a string.
     * Handle null strings by returning null.
     * 
     * Examples:
     * reverseString("hello") → "olleh"
     * reverseString("Java") → "avaJ"
     * reverseString("") → ""
     * reverseString("a") → "a"
     * 
     * @param str Input string to reverse
     * @return Reversed string
     */
    public String reverseString(String str) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Check if a string is a palindrome (reads the same forwards and backwards).
     * Case-insensitive comparison.
     * Handle null strings by returning false.
     * 
     * Examples:
     * isPalindrome("racecar") → true
     * isPalindrome("A man a plan a canal Panama") → true (ignoring spaces and case)
     * isPalindrome("hello") → false
     * isPalindrome("") → true
     * 
     * @param str Input string to check
     * @return True if string is a palindrome
     */
    public boolean isPalindrome(String str) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Capitalize the first letter of each word in a string.
     * Words are separated by spaces.
     * Handle null strings by returning null.
     * 
     * Examples:
     * capitalizeWords("hello world") → "Hello World"
     * capitalizeWords("java programming") → "Java Programming"
     * capitalizeWords("a") → "A"
     * capitalizeWords("") → ""
     * 
     * @param str Input string to capitalize
     * @return String with first letter of each word capitalized
     */
    public String capitalizeWords(String str) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Build a string efficiently using StringBuilder.
     * Concatenate all strings in the array with the given separator.
     * This demonstrates the preferred way to build strings in Java.
     * 
     * Examples:
     * buildString(["Hello", "World", "Java"], " ") → "Hello World Java"
     * buildString(["a", "b", "c"], "-") → "a-b-c"
     * buildString([], ",") → ""
     * 
     * @param parts Array of strings to concatenate
     * @param separator Separator to use between parts
     * @return Concatenated string built with StringBuilder
     */
    public String buildString(String[] parts, String separator) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Extract all words from a string (separated by spaces).
     * Remove empty strings from the result.
     * Handle multiple consecutive spaces.
     * 
     * Examples:
     * extractWords("hello world java") → ["hello", "world", "java"]
     * extractWords("  multiple   spaces  ") → ["multiple", "spaces"]
     * extractWords("") → []
     * extractWords("single") → ["single"]
     * 
     * @param str Input string to extract words from
     * @return List of words extracted from the string
     */
    public List<String> extractWords(String str) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Check if a string represents a valid integer.
     * Handle negative numbers and leading/trailing whitespace.
     * 
     * Examples:
     * isValidInteger("123") → true
     * isValidInteger("-456") → true
     * isValidInteger("  789  ") → true
     * isValidInteger("12.34") → false
     * isValidInteger("abc") → false
     * isValidInteger("") → false
     * 
     * @param str String to check
     * @return True if string represents a valid integer
     */
    public boolean isValidInteger(String str) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }

    /**
     * Remove all non-alphabetic characters from a string.
     * Keep only letters (a-z, A-Z).
     * 
     * Examples:
     * removeNonAlphabetic("Hello123World!") → "HelloWorld"
     * removeNonAlphabetic("Java2024") → "Java"
     * removeNonAlphabetic("123!@#") → ""
     * removeNonAlphabetic("") → ""
     * 
     * @param str Input string to clean
     * @return String with only alphabetic characters
     */
    public String removeNonAlphabetic(String str) {
        throw new UnsupportedOperationException("Method not implemented yet");
    }
}
